savant/deepstream/buffer_processor.py

```diff
@@ -105,13 +106,33 @@ def prepare_input(self, buffer: Gst.Buffer):
 
             frame_meta = metadata_get_frame_meta(source_id, frame_idx, frame_pts)
 
+            # full frame primary object by default
+            primary_bbox = BBox(
+                x_center=self._frame_params.width / 2,
+                y_center=self._frame_params.height / 2,
+                width=self._frame_params.width,
+                height=self._frame_params.height,
+            )
+
             # add external objects to nvds meta
             for obj_meta in frame_meta.metadata['objects']:
                 obj_key = self._model_object_registry.model_object_key(
                     obj_meta['model_name'], obj_meta['label']
                 )
-                # skip frame, will be added with proper width/height
-                if obj_key == 'frame':
+                # skip primary object for now, will be added later
+                if obj_key == PRIMARY_OBJECT_LABEL:
+                    bbox = (
+                        obj_meta['bbox']['xc'],
+                        obj_meta['bbox']['yc'],
+                        obj_meta['bbox']['width'],
+                        obj_meta['bbox']['height'],
+                    )
+                    # if not a full frame then correct primary object
+                    if bbox != (0.5, 0.5, 1, 1):
```

Comment:
I suggest adding a warning log message because if the first object is not equal to the whole frame, it is abnormal.

Reply:
Why is it abnormal? For example, I want to remove a bottom area from processing (there is the camera name and date), and I want to do it for cascaded pipelines. Why not?

Reply:
You are right. Now the frame and the primary RoI object are the same and it's a bit confusing. With this code, I'm just providing a way to change the RoI. I think we need to decide how to deal with it, but not in this task. I have some thoughts, which we'll discuss in a meeting.